src/Libraries/Database/Idiorm/Paginator.php

```diff
@@ -0,0 +1,301 @@
+	 * @throws DatabaseException
+	 */
+	public function __construct($idiormDbal, int $per_page, int $page = 1)
+	{
+		/** @var IdiormDbal $idiormDbal */
+		$this->setTotal($idiormDbal);
+		$this->dbal = $idiormDbal;
+		$this->dbal->limit($per_page)->offset($per_page * ($page - 1));
+		$this->data = $this->dbal->getOrmModel()->find_many();
+		$this->per_page = $per_page;
+		$this->page = $page;
+	}
+
+	/**
+	 * @return int
+	 */
+	public function currentPageNumber(): int
+	{
+		return $this->page;
+	}
+
+	/**
+	 * @param bool $withBaseUrl
+	 * @return string|null
+	 */
+	public function currentPageLink(bool $withBaseUrl = false): ?string
+	{
+		$current = null;
+		if (!empty($this->page)) {
+			$current = $this->getUri($withBaseUrl) . 'per_page=' . $this->per_page . '&page=' . $this->page;
+		}
+		return $current;
+	}
+
+	/**
+	 * @return int|null
+	 */
+	public function previousPageNumber(): ?int
+	{
+		$previous = null;
+		if ($this->page > 1) {
+			$previous = $this->page - 1;
+		} elseif ($this->page == 1) {
+			$previous = $this->page;
+		}
+
+		return $previous;
+	}
+
+	/**
+	 * @param bool $withBaseUrl
+	 * @return string|null
+	 */
+	public function previousPageLink(bool $withBaseUrl = false): ?string
+	{
+		$previous = null;
+		if (!empty($this->previousPageNumber())) {
+			$previous = $this->getUri($withBaseUrl) . 'per_page=' . $this->per_page . '&page=' . $this->previousPageNumber();
+		}
+		return $previous;
+	}
+
+	/**
+	 * @return int|null
+	 */
+	public function nextPageNumber(): ?int
+	{
+		$next = null;
+		if ($this->page < $this->lastPageNumber()) {
+			$next = $this->page + 1;
+		} elseif ($this->page == $this->lastPageNumber()) {
+			$next = $this->page;
+		}
+		return $next;
+	}
+
+	/**
+	 * @param bool $withBaseUrl
+	 * @return string|null
+	 */
+	public function nextPageLink(bool $withBaseUrl = false): ?string
+	{
+		$next = null;
+		if (!empty($this->nextPageNumber())) {
+			$next = $this->getUri($withBaseUrl) . 'per_page=' . $this->per_page . '&page=' . $this->nextPageNumber();
+		}
+		return $next;
+	}
+
+	/**
+	 * @param bool $withBaseUrl
+	 * @return string|null
+	 */
+	public function firstPageLink(bool $withBaseUrl = false): ?string
+	{
+		return $this->getUri($withBaseUrl) . 'per_page=' . $this->per_page . '&page=1';
+	}
+
+	/**
+	 * @return int
+	 */
+	public function lastPageNumber(): int
+	{
+		return (int)ceil($this->total() / $this->per_page);
+	}
+
+	/**
+	 * @param bool $withBaseUrl
+	 * @return string|null
+	 */
+	public function lastPageLink(bool $withBaseUrl = false): ?string
+	{
+		$last = null;
+		if (!empty($this->lastPageNumber())) {
+			$last = $this->getUri($withBaseUrl) . 'per_page=' . $this->per_page . '&page=' . $this->lastPageNumber();
+		}
+		return $last;
+	}
+
+	/**
+	 * @return mixed
+	 */
+	public function firstItem()
+	{
+		return $this->data[array_key_first($this->data)];
+	}
+
+	/**
+	 * @return mixed
+	 */
+	public function lastItem()
+	{
+		return $this->data[array_key_last($this->data)];
```

Comment:
It seems like `$this->data` can also be of type `IdiormResultSet`; however, parameter `$array` of `array_key_last()` does only seem to accept array, maybe add an additional type check?